### PR TITLE
Order3

### PR DESCRIPTION
**Pull request by jeremyneveu**

### CODE REVIEW

File: spectractor/extractor/extractor.py [hunk 1]
```diff
@@ -416,9 +430,9 @@ def simulate(self, *params):
         # linear regression for the amplitude parameters
         # prepare the vectors
         self.p = np.asarray(params)
-        A2, D2CCD, dx0, dy0, angle, B, rot, pressure, temperature, airmass, *poly_params = params
-        poly_params_order1 = poly_params[:len(poly_params)//2]
-        poly_params_order2 = poly_params[len(poly_params)//2:]
+        A2, D2CCD, dx0, dy0, angle, B, rot, pressure, temperature, airmass, *poly_params_all = params
+        poly_params = poly_params_all[:len(poly_params_all)//2]
+        poly_params_next_order = poly_params_all[len(poly_params_all)//2:]
```
Comment (mfisherlevine) [inline spectractor/extractor/extractor.py:435]: Seeing integer division here scares me a little bit, but I guess if that's guaranteed to always be even by definition then that's fine? (Like if it's some length things that put in there twice, for example). Maybe sort sort of assert/check just after to check nothing went wrong there somehow though?

File: spectractor/extractor/extractor.py [hunk 2]
```diff
@@ -436,38 +450,51 @@ def simulate(self, *params):
...
+            profile_params_next_order[:, 1] = dispersion_law_next_order.real + self.spectrum.spectrogram_x0
+            profile_params_next_order[:, 2] += dispersion_law_next_order.imag - self.bgd_width
+
+            psf_cube_next_order = A2 * self.spectrum.chromatic_psf.build_psf_cube(self.pixels, profile_params_next_order,
+                                                                                 fwhmx_clip=3 * parameters.PSF_FWHM_CLIP,
+                                                                                 fwhmy_clip=parameters.PSF_FWHM_CLIP,
+                                                                                 dtype="float32", mask=self.psf_cube_masked_next_order)
+            psf_cube += psf_cube_next_order
+            if self.tr_ratio_next_next_order is not None:
+                self.my_logger.warning("gogogo")
```
Comment (mfisherlevine) [inline spectractor/extractor/extractor.py:483]: Was this message supposed to live into the production code?

File: spectractor/simulation/simulator.py [hunk 3]
```diff
@@ -431,11 +430,14 @@ def simulate(self, A1=1.0, A2=0., aerosols=0.05, ozone=300, pwv=5, D=parameters.
...
         # Add order 2
-        if A2 > 0.:
+        if A2 > 0. and self.disperser.ratio_order_2over1 is not None:
             spectrum_order2 = self.disperser.ratio_order_2over1(self.lambdas) * spectrum
             spectrum_order2_err = self.disperser.ratio_order_2over1(self.lambdas) * spectrum_err
             if np.any(np.isnan(spectrum_order2)):
                 spectrum_order2[np.isnan(spectrum_order2)] = 0.
+            dispersion_law_order2 = self.compute_dispersion_in_spectrogram(self.lambdas, shift_x, shift_y, angle,
+                                                                           niter=5, with_adr=True,
+                                                                           order=self.order + np.sign(self.order))
```
Comment (mfisherlevine) [inline spectractor/simulation/simulator.py:440]: I'm not quite sure what the  `np.sign(self.order)` part is going, but adding it feels strange, as opposed to multiplying by it, given that it's +/-1. Are you sure that's right? (There's a similar line somewhere above too)

File: spectractor/extractor/extractor.py (hunk 1 above)
Comment (jeremyneveu) [reply spectractor/extractor/extractor.py:435]: Ok, I add a check with a raise ValueError

File: spectractor/simulation/simulator.py (hunk 3 above)
Comment (jeremyneveu) [reply spectractor/simulation/simulator.py:440]: Yes I am sure: if specified order is negative, then next order is order-1, and second next order is order-2, so you take the sign of the given order (np.sign returns -1)  and add it.

Comment (mfisherlevine) [reply spectractor/simulation/simulator.py:443]: Haha, OK, that makes sense, thanks 🙂

**Pull request merged**